prymer/api/oligo.py

```diff
@@ -0,0 +1,207 @@
+>>> oligo_span = Span(refname="chr1", start=1, end=20)
+>>> oligo = Oligo(tm=70.0, penalty=-123.0, span=oligo_span, bases="AGCT" * 5)
+>>> oligo.longest_hp_length()
+1
+>>> oligo.length
+20
+>>> oligo.name is None
+True
+>>> oligo = Oligo(tm=70.0, penalty=-123.0, span=oligo_span, bases="GACGG"*4)
+>>> oligo.longest_hp_length()
+3
+>>> oligo.untailed_length()
+20
+>>> oligo.tailed_length()
+20
+>>> primer = oligo.with_tail(tail="GATTACA")
+>>> primer.untailed_length()
+20
+>>> primer.tailed_length()
+27
+>>> primer = primer.with_name(name="fwd_primer")
+>>> primer.name
+'fwd_primer'
+
+```
+
+Oligos may also be written to a file and subsequently read back in, as the `Oligo` class is an
+`fgpyo` `Metric` class:
+
+```python
+>>> from pathlib import Path
+>>> left_span = Span(refname="chr1", start=1, end=20)
+>>> left = Oligo(tm=70.0, penalty=-123.0, span=left_span, bases="G"*20)
+>>> right_span = Span(refname="chr1", start=101, end=120)
+>>> right = Oligo(tm=70.0, penalty=-123.0, span=right_span, bases="T"*20)
+>>> path = Path("/tmp/path/to/primers.txt")
+>>> Oligo.write(path, left, right)  # doctest: +SKIP
+>>> primers = Oligo.read(path)  # doctest: +SKIP
+>>> list(primers)  # doctest: +SKIP
+[
+    Oligo(tm=70.0, penalty=-123.0, span=amplicon_span, bases="G"*20),
+    Oligo(tm=70.0, penalty=-123.0, span=amplicon_span, bases="T"*20)
+]
+
+```
+"""
+
+from dataclasses import dataclass
+from dataclasses import replace
+from typing import Any
+from typing import Callable
+from typing import Dict
+from typing import Optional
+
+from fgpyo.fasta.sequence_dictionary import SequenceDictionary
+from fgpyo.sequence import longest_dinucleotide_run_length
+from fgpyo.sequence import longest_homopolymer_length
+from fgpyo.util.metric import Metric
+
+from prymer.api.oligo_like import MISSING_BASES_STRING
+from prymer.api.oligo_like import OligoLike
+from prymer.api.span import Span
+
+
+@dataclass(frozen=True, init=True, kw_only=True, slots=True)
+class Oligo(OligoLike, Metric["Oligo"]):
+    """Stores the properties of the designed oligo.
+
+    Oligos can include both single primer and internal probe designs. The penalty score of the
+    design is emitted by Primer3 and controlled by the corresponding design parameters.
+    The penalty for a primer is set by the combination of `PrimerAndAmpliconParameters` and
+    `PrimerWeights`, whereas a probe penalty is set by `ProbeParameters` and `ProbeWeights`.
+
+    Attributes:
+        tm: the calculated melting temperature of the oligo
+        penalty: the penalty or score for the oligo
+        span: the mapping of the primer to the genome
+        bases: the base sequence of the oligo (excluding any tail)
+        tail: an optional tail sequence to put on the 5' end of the primer
+        name: an optional name to use for the primer
+
+    """
+
+    tm: float
+    penalty: float
+    span: Span
+    bases: Optional[str] = None
+    tail: Optional[str] = None
+
+    def __post_init__(self) -> None:
```

Comment:
all of these methods derive from the original `Primer` class and are unchanged, just ported from `primer.py` to `oligo.py`